File: Application/Sources/UI/Controllers/PageContainerViewController.m
//
//  Copyright (c) SRG SSR. All rights reserved.
//
//  License information is available from the LICENSE file.
//

#import "PageContainerViewController.h"

#import "PlayLogger.h"
#import "UIColor+PlaySRG.h"
#import "UIViewController+PlaySRG.h"
#import "UIVisualEffectView+PlaySRG.h"

#import "MaterialTabs.h"
@import SRGAppearance;

@interface PageContainerViewController () <MDCTabBarDelegate>

@property (nonatomic) UIPageViewController *pageViewController;

@property (nonatomic) NSArray<UIViewController *> *viewControllers;
@property (nonatomic) NSUInteger initialPage;

@property (nonatomic, weak) MDCTabBar *tabBar;
@property (nonatomic, weak) UIVisualEffectView *blurView;

@end

@implementation PageContainerViewController

#pragma mark Object lifecycle

- (instancetype)initWithViewControllers:(NSArray<UIViewController *> *)viewControllers initialPage:(NSUInteger)initialPage
{
    NSAssert(viewControllers.count != 0, @"At least one view controller is required");
    
    if (self = [super init]) {
        self.viewControllers = viewControllers;
        
        if (initialPage >= viewControllers.count) {
            PlayLogWarning(@"pageViewController", @"Invalid page. Fixed to 0");
            initialPage = 0;
        }
        self.initialPage = initialPage;
        
        UIPageViewController *pageViewController = [[UIPageViewController alloc] initWithTransitionStyle:UIPageViewControllerTransitionStyleScroll
                                                                                   navigationOrientation:UIPageViewControllerNavigationOrientationHorizontal
                                                                                                 options:@{ UIPageViewControllerOptionInterPageSpacingKey : @100.f }];
        pageViewController.delegate = self;
        
        // Only allow scrolling if several pages are available
        if (viewControllers.count > 1) {
            pageViewController.dataSource = self;
        }
        
        self.pageViewController = pageViewController;
        [self addChildViewController:pageViewController];
    }
    return self;
}

- (instancetype)initWithViewControllers:(NSArray<UIViewController *> *)viewControllers
{
    return [self initWithViewControllers:viewControllers initialPage:0];
}

#pragma mark View lifecycle

- (void)loadView
{
    self.view = [[UIView alloc] initWithFrame:UIScreen.mainScreen.bounds];
    self.view.backgroundColor = UIColor.srg_gray16Color;
    
    UIView *pageView = self.pageViewController.view;
    [self.view insertSubview:pageView atIndex:0];
    
    pageView.translatesAutoresizingMaskIntoConstraints = NO;
    [NSLayoutConstraint activateConstraints:@[
        [pageView.topAnchor constraintEqualToAnchor:self.view.topAnchor],
        [pageView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor],
        [pageView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [pageView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor]
    ]];
    
    [self.pageViewController didMoveToParentViewController:self];
    
    UIVisualEffectView *blurView = UIVisualEffectView.play_blurView;
    [self.view addSubview:blurView];
    self.blurView = blurView;
    
    blurView.translatesAutoresizingMaskIntoConstraints = NO;
    [NSLayoutConstraint activateConstraints:@[
        [blurView.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor],
        [blurView.heightAnchor constraintEqualToConstant:60.f],
        [blurView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [blurView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor]
    ]];
    
    __block BOOL hasImage = NO;
    
    NSMutableArray<UITabBarItem *> *tabBarItems = [NSMutableArray array];
    [self.viewControllers enumerateObjectsUsingBlock:^(UIViewController * _Nonnull viewController, NSUInteger idx, BOOL * _Nonnull stop) {
        UITabBarItem *tabBarItem = viewController.tabBarItem;
        if (tabBarItem.image) {
            hasImage = YES;
        }
        [tabBarItems addObject:tabBarItem];
    }];
    
    MDCTabBar *tabBar = [[MDCTabBar alloc] init];
    tabBar.itemAppearance = hasImage ? MDCTabBarItemAppearanceImages : MDCTabBarItemAppearanceTitles;
    tabBar.alignment = MDCTabBarAlignmentCenter;
    tabBar.delegate = self;
    tabBar.items = tabBarItems.copy;
    
    tabBar.tintColor = UIColor.whiteColor;
    tabBar.unselectedItemTintColor = UIColor.srg_gray96Color;
    tabBar.selectedItemTintColor = UIColor.whiteColor;
        
    // Use ripple effect without color, so that there is no Material-like highlighting (we are NOT adopting Material)
    tabBar.enableRippleBehavior = YES;
    tabBar.rippleColor = UIColor.clearColor;
    
    [blurView.contentView addSubview:tabBar];
    self.tabBar = tabBar;
    
    tabBar.translatesAutoresizingMaskIntoConstraints = NO;
    [NSLayoutConstraint activateConstraints:@[
        [tabBar.topAnchor constraintEqualToAnchor:blurView.contentView.topAnchor],
        [tabBar.bottomAnchor constraintEqualToAnchor:blurView.contentView.bottomAnchor],
        [tabBar.leadingAnchor constraintEqualToAnchor:blurView.contentView.leadingAnchor],
        [tabBar.trailingAnchor constraintEqualToAnchor:blurView.contentView.trailingAnchor]
    ]];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(pageContainerViewController_contentSizeCategoryDidChange:)
                                                 name:UIContentSizeCategoryDidChangeNotification
                                               object:nil];
    
    [self updateFonts];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.tabBar.selectedItem = self.tabBar.items[self.initialPage];
    
    UIViewController *initialViewController = self.viewControllers[self.initialPage];
    [self.pageViewController setViewControllers:@[initialViewController] direction:UIPageViewControllerNavigationDirectionForward animated:NO completion:nil];
    [self didDisplayViewController:initialViewController animated:NO];
}

- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator
{
    [super viewWillTransitionToSize:size withTransitionCoordinator:coordinator];
    
    [coordinator animateAlongsideTransition:^(id<UIViewControllerTransitionCoordinatorContext>  _Nonnull context) {
        // Force a refresh of the tab bar so that the alignment is correct after rotation
        self.tabBar.alignment = MDCTabBarAlignmentLeading;
        self.tabBar.alignment = MDCTabBarAlignmentCenter;
    } completion:nil];
}

#pragma mark Rotation

- (UIInterfaceOrientationMask)supportedInterfaceOrientations
{
    return [super supportedInterfaceOrientations] & UIViewController.play_supportedInterfaceOrientations;
}

#pragma mark Status bar

- (UIStatusBarStyle)preferredStatusBarStyle
{
    return UIStatusBarStyleLightContent;
}

#pragma mark Accessibility

- (BOOL)accessibilityPerformEscape
{
    if (self.navigationController.viewControllers.count > 1) {
        [self.navigationController popViewControllerAnimated:YES];
        return YES;
    }
    else if (self.presentingViewController) {
        [self dismissViewControllerAnimated:YES completion:nil];
        return YES;
    }
    else {
        return NO;
    }
}

#pragma mark Actions

- (BOOL)switchToIndex:(NSUInteger)index animated:(BOOL)animated
{
    if (! [self displayPageAtIndex:index animated:animated]) {
        return NO;
    }
    
    if (self.tabBar) {
        [self.tabBar setSelectedItem:self.tabBar.items[index] animated:animated];
    }
    else {
        self.initialPage = index;
    }
    return YES;
}

#pragma mark Display

- (BOOL)displayPageAtIndex:(NSUInteger)index animated:(BOOL)animated
{
    if (index >= self.viewControllers.count) {
        return NO;
    }
    
    UIViewController *currentViewController = self.pageViewController.viewControllers.firstObject;
    NSUInteger currentIndex = [self.viewControllers indexOfObject:currentViewController];
    UIPageViewControllerNavigationDirection direction = (currentIndex < index) ? UIPageViewControllerNavigationDirectionForward : UIPageViewControllerNavigationDirectionReverse;
    
    UIViewController *newViewController = self.viewControllers[index];
    [self.pageViewController setViewControllers:@[newViewController] direction:direction animated:animated completion:nil];
    
    [self didDisplayViewController:newViewController animated:animated];
    return YES;
}

#pragma mark UI

- (void)updateFonts
{
    UIFont *tabBarFont = [SRGFont fontWithStyle:SRGFontStyleBody];
    self.tabBar.unselectedItemTitleFont = tabBarFont;
    self.tabBar.selectedItemTitleFont = tabBarFont;
}

#pragma mark Stubs

- (void)didDisplayViewController:(UIViewController *)viewController animated:(BOOL)animated
{}

#pragma mark ContainerContentInsets protocol

- (UIEdgeInsets)play_additionalContentInsets
{
    return UIEdgeInsetsMake(CGRectGetHeight(self.blurView.frame), 0.f, 0.f, 0.f);
}

#pragma mark MDCTabBarDelegate protocol

- (void)tabBar:(MDCTabBar *)tabBar didSelectItem:(UITabBarItem *)item
{
    NSUInteger index = [tabBar.items indexOfObject:item];
    [self displayPageAtIndex:index animated:YES];
}

#pragma mark SRGAnalyticsContainerViewTracking protocol

- (NSArray<UIViewController *> *)srg_activeChildViewControllers
{
    return self.pageViewController ? @[self.pageViewController] : @[];
}

#pragma mark TabBarActionable protocol

- (void)performActiveTabActionAnimated:(BOOL)animated
{
    UIViewController *currentViewController = self.pageViewController.viewControllers.firstObject;
    if ([currentViewController conformsToProtocol:@protocol(TabBarActionable)]) {
        UIViewController<TabBarActionable> *actionableCurrentViewController = (UIViewController<TabBarActionable> *)currentViewController;
        [actionableCurrentViewController performActiveTabActionAnimated:animated];
    }
}

#pragma mark UIPageViewControllerDataSource protocol

- (UIViewController *)pageViewController:(UIPageViewController *)pageViewController viewControllerBeforeViewController:(UIViewController *)viewController
{
    NSUInteger currentIndex = [self.viewControllers indexOfObject:viewController];
    if (currentIndex > 0) {
        return self.viewControllers[currentIndex - 1];
    }
    else {
        return nil;
    }
}

- (UIViewController *)pageViewController:(UIPageViewController *)pageViewController viewControllerAfterViewController:(UIViewController *)viewController
{
    NSUInteger currentIndex = [self.viewControllers indexOfObject:viewController];
    if (currentIndex < self.viewControllers.count - 1) {
        return self.viewControllers[currentIndex + 1];
    }
    else {
        return nil;
    }
}

#pragma mark UIPageViewControllerDelegate protocol

- (void)pageViewController:(UIPageViewController *)pageViewController didFinishAnimating:(BOOL)finished previousViewControllers:(NSArray<UIViewController *> *)previousViewControllers transitionCompleted:(BOOL)completed
{
    if (completed) {
        UIViewController *newViewController = pageViewController.viewControllers.firstObject;
        NSUInteger currentIndex = [self.viewControllers indexOfObject:newViewController];
        [self.tabBar setSelectedItem:self.tabBar.items[currentIndex] animated:YES];;
        [self didDisplayViewController:newViewController animated:YES];
    }
}

#pragma mark Notifications

- (void)pageContainerViewController_contentSizeCategoryDidChange:(NSNotification *)notification
{
    [self updateFonts];
}

@end

@implementation UIViewController (PageContainerViewController)

#pragma mark Getters and setters

- (PageContainerViewController *)play_pageContainerViewController
{
    UIViewController *parentViewController = self.parentViewController.parentViewController;
    if ([parentViewController isKindOfClass:PageContainerViewController.class]) {
        return (PageContainerViewController *)parentViewController;
    }
    else {
        return nil;
    }
}

@end

File: Application/Sources/UI/Controllers/PageContainerViewController.swift
//
//  Copyright (c) SRG SSR. All rights reserved.
//
//  License information is available from the LICENSE file.
//

import Combine
import Pageboy
import SRGAppearance
import Tabman
import UIKit

class PageContainerViewController: UIViewController {
    let viewControllers: [UIViewController]
    private let additionalViewControllers: [UIViewController]

    private var tabContainerViewController: TabContainerViewController
    private(set) var initialPage: Int
    private let tabBarItems: [TMBarItem]
    private weak var tabBarTopConstraint: NSLayoutConstraint?
    private weak var blurView: UIVisualEffectView?
    private var cancellables: Set<AnyCancellable> = []
    private var satelliteRadioChannels: [RadioChannel] = []
    private var cancellable: AnyCancellable?

    init(viewControllers: [UIViewController], additionalViewControllers: [UIViewController], satelliteRadioChannels: [RadioChannel], initialPage: Int) {
        assert(!viewControllers.isEmpty, "At least one view controller is required")

        self.viewControllers = viewControllers
        self.additionalViewControllers = additionalViewControllers
        self.satelliteRadioChannels = satelliteRadioChannels

        if initialPage >= 0, initialPage < viewControllers.count {
            self.initialPage = initialPage
        } else {
            PlayLogWarning(category: "pageViewController", message: "Invalid page. Fixed to 0")
            self.initialPage = 0
        }

        tabBarItems = viewControllers.appending(contentsOf: additionalViewControllers).map {
            if let tabBarItem = $0.tabBarItem, let image = tabBarItem.image {
                let item = TMBarItem(image: image)
                item.accessibilityLabel = tabBarItem.title ?? $0.title
                return item
            } else {
                let item = TMBarItem(title: $0.title ?? "")
                item.accessibilityLabel = $0.title
                return item
            }
        }

        tabContainerViewController = TabContainerViewController()

        super.init(nibName: nil, bundle: nil)

        addChild(tabContainerViewController)
    }

    convenience init(viewControllers: [UIViewController], additionalViewControllers: [UIViewController], satelliteRadioChannels: [RadioChannel]) {
        self.init(viewControllers: viewControllers, additionalViewControllers: additionalViewControllers, satelliteRadioChannels: satelliteRadioChannels, initialPage: 0)
    }

    @available(*, unavailable)
    required init?(coder _: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    private func configureBarView() {
        let blurView = UIVisualEffectView.play_blurView
        self.blurView = blurView

        let barView = TMBarView<TMHorizontalBarLayout, TMTabItemBarButton, TMLineBarIndicator>()
        barView.backgroundView.style = .custom(view: blurView)
        barView.layout.alignment = .centerDistributed
        barView.indicator.tintColor = .white

        var buttonIndex = 0
        barView.buttons.customize { [weak self] button in
            guard let self else { return }
            button.tag = buttonIndex
            button.imageContentMode = .center
            button.addTarget(self, action: #selector(tabDidChange(_:)), for: .touchUpInside)
            buttonIndex += 1
        }
        tabContainerViewController.addBar(barView, dataSource: self, at: .top)
    }

    override func loadView() {
        view = UIView(frame: UIScreen.main.bounds)
        view.backgroundColor = .srgGray16
        view = view
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        configureBarView()
        if let tabContainerView = tabContainerViewController.view {
            view.insertSubview(tabContainerView, at: 0)

            tabContainerView.translatesAutoresizingMaskIntoConstraints = false
            let tabBarTopConstraint = tabContainerView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor)
            NSLayoutConstraint.activate([
                tabBarTopConstraint,
                tabContainerView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
                tabContainerView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
                tabContainerView.trailingAnchor.constraint(equalTo: view.trailingAnchor)
            ])
            self.tabBarTopConstraint = tabBarTopConstraint
        }
        tabContainerViewController.didMove(toParent: self)

        tabContainerViewController.dataSource = self
        didDisplayViewController(tabContainerViewController.currentViewController, animated: false)

        tabContainerViewController
            .updateSignal()
            .debounce(for: 0.1, scheduler: DispatchQueue.main)
            .sink { [weak self] _ in
                guard let self else { return }
                didDisplayViewController(tabContainerViewController.currentViewController, animated: false)
            }
            .store(in: &cancellables)
    }

    override var preferredStatusBarStyle: UIStatusBarStyle {
        .lightContent
    }

    override func accessibilityPerformEscape() -> Bool {
        if let navigationController, navigationController.viewControllers.count > 1 {
            navigationController.popViewController(animated: true)
            return true
        } else if presentingViewController != nil {
            dismiss(animated: true, completion: nil)
            return true
        } else {
            return false
        }
    }

    func switchToIndex(_ index: Int, animated: Bool) -> Bool {
        guard index < viewControllers.count else { return false }

        if isViewLoaded {
            tabContainerViewController.scrollToPage(.at(index: index), animated: animated)
            didDisplayViewController(tabContainerViewController.currentViewController, animated: animated)
            return true
        } else {
            initialPage = index
            return true
        }
    }

    func didDisplayViewController(_: UIViewController?, animated _: Bool) {
        play_setNeedsScrollableViewUpdate()
    }
}

// MARK: - Protocols

extension PageContainerViewController: ContainerContentInsets {
    var play_additionalContentInsets: UIEdgeInsets {
        UIEdgeInsets(top: tabContainerViewController.barInsets.top, left: 0.0, bottom: 0.0, right: 0.0)
    }
}

extension PageContainerViewController: Oriented {
    var play_supportedInterfaceOrientations: UIInterfaceOrientationMask {
        .all
    }

    var play_orientingChildViewControllers: [UIViewController] {
        viewControllers
    }
}

extension PageContainerViewController: ScrollableContentContainer {
    var play_scrollableChildViewController: UIViewController? {
        tabContainerViewController.currentViewController
    }

    func play_contentOffsetDidChange(inScrollableView scrollView: UIScrollView) {
        let adjustedOffset = scrollView.contentOffset.y + scrollView.adjustedContentInset.top
        tabBarTopConstraint?.constant = max(-adjustedOffset, 0.0)
    }
}

extension PageContainerViewController: SRGAnalyticsContainerViewTracking {
    var srg_activeChildViewControllers: [UIViewController] {
        [tabContainerViewController]
    }
}

extension PageContainerViewController: TabBarActionable {
    func performActiveTabAction(animated: Bool) {
        if let currentViewController = tabContainerViewController.currentViewController,
           let actionableCurrentViewController = currentViewController as? TabBarActionable {
            actionableCurrentViewController.performActiveTabAction(animated: animated)
        }
    }
}

extension UIViewController {
    var play_pageContainerViewController: PageContainerViewController? {
        var parentViewController = parent
        while let viewController = parentViewController {
            if let pageContainerViewController = viewController as? PageContainerViewController {
                return pageContainerViewController
            }
            parentViewController = viewController.parent
        }
        return nil
    }
}

extension PageContainerViewController: PageboyViewControllerDataSource, TMBarDataSource {
    func numberOfViewControllers(in _: Pageboy.PageboyViewController) -> Int {
        viewControllers.count + additionalViewControllers.count
    }

    func viewController(for _: Pageboy.PageboyViewController, at index: Pageboy.PageboyViewController.PageIndex) -> UIViewController? {
        if index < viewControllers.count {
            viewControllers[index]
        } else {
            nil
        }
    }

    func defaultPage(for _: Pageboy.PageboyViewController) -> Pageboy.PageboyViewController.Page? {
        .at(index: initialPage)
    }

    func barItem(for _: any Tabman.TMBar, at index: Int) -> any Tabman.TMBarItemable {
        tabBarItems[index]
    }
}

// MARK: Swiss Satellite Radio

extension PageContainerViewController {
    func srgMedia(for radioChannel: RadioChannel) -> AnyPublisher<SRGMedia, Error> {
        SRGDataProvider.current!.regionalizedRadioLivestreams(for: ApplicationConfiguration.shared.vendor, contentProviders: .swissSatelliteRadio)
            .compactMap { $0.first { $0.uid == radioChannel.uid } }
            .eraseToAnyPublisher()
    }

    @objc private func tabDidChange(_ sender: TMTabItemBarButton) {
        if sender.tag >= viewControllers.count {
            cancellable = srgMedia(for: satelliteRadioChannels[sender.tag - viewControllers.count])
                .receive(on: DispatchQueue.main)
                .sink(
                    receiveCompletion: { _ in },
                    receiveValue: { [weak self] srgMedia in
                        self?.play_presentMediaPlayer(with: srgMedia, position: nil, airPlaySuggestions: true, fromPushNotification: false, animated: true, completion: nil)
                    }
                )
        }
    }
}
